### PR TITLE
Implement should_override_forkchoice_update

### PR DESCRIPTION
**Pull request by potuz**
Part of implementing honest validators orphaning late blocks.

### CODE REVIEW

**Review by terencechain — COMMENTED** [review]
Looks reasonable to me. This is blocked by #11964. A little hard to review until that is merged

File: beacon-chain/forkchoice/doubly-linked-tree/proposer_boost.go
Comment (nisdas) [inline beacon-chain/forkchoice/doubly-linked-tree/proposer_boost.go:23]: its been a while since I looked at forkchoice code, but this mutex ordering won't cause any issues, correct ? I know that previously we had many deadlocks due to how the different locks/rlocks were acquired and released.

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go
Comment (nisdas) [inline beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:56]: again, same question on mutex ordering

**Review by terencechain — CHANGES_REQUESTED** [review]
Can we add a feature flag. I'm with it either inverted or default

**Comment by potuz** [comment]
> Can we add a feature flag. I'm with it either inverted or default

How? this doesn't have any runtime changes a feature flag to gate what?

File: beacon-chain/forkchoice/doubly-linked-tree/proposer_boost.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/proposer_boost.go:23]: Discussed offline

File: beacon-chain/forkchoice/doubly-linked-tree/node.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/node.go:21]: any thoughts on putting these as flags with default to 4 / 10 or we simply do not want ppl to change these

File: beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go:107]: Let's just remove `activeBalance` from `UnrealizedCheckpoints`. It's cleaner

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:16]: These format look messed up

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:51]: Dont like these silent errors, can we add any log?

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:59]: Shouldn't this be `>=` instead.
Current epoch: 4
Justified epoch: 3
Target epoch: 2
That's a happy case no?

Comment (terencechain) [reply beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:59]: Ignore. I calculated it wrong

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:75]:
```suggestion
	// Only orphan a block if the head LMD vote is weak
```

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:27]: I think we are missing this:
```
    # Only re-org if we are proposing on-time.
    time_into_slot = (store.time - store.genesis_time) % SECONDS_PER_SLOT
    proposing_on_time = time_into_slot <= SECONDS_PER_SLOT // INTERVALS_PER_SLOT // 2
```

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks_test.go
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks_test.go:11]: We are missing an important test case where head arrives early

Comment (terencechain) [reply beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks_test.go:11]:
```go
	if early {
		return
	}
```

File: beacon-chain/forkchoice/doubly-linked-tree/node.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/node.go:21]: I rather users not mess with these since that can cause more trouble than what we are trying to fix.

File: beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go:107]: nice one!

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks.go:27]: Yeah, that can't really happen, we will only call this function after receiving a block and we are proposing during the next slot. The spec implementation allows you to reorg a late block that was supposed to be ready during N and that arrives during N+1. We will always reorg such blocks anyway.

File: beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks_test.go
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/reorg_late_blocks_test.go:11]: nice!

**Review by terencechain — APPROVED** [review]
Lgtm. Crosschecked against spec coverages and unit test coverages. This is not used in run time yet

**Pull request merged**
